File: python-wrapper/examples/multimodal_output.py
import asyncio
import base64
import os
from pathlib import Path
from ferrant import Agent

async def main() -> None:
    agent = Agent.openai("gpt-audio-1.5", os.environ["OPENAI_API_KEY"],
        modalities=["text", "audio"], audio_format="wav", audio_voice="alloy")
    response = await agent.run_multimodal([
        {"type": "text", "text": "Give Rust a friendly welcome."}
    ])
    print("Transcript:", response["content"])
    for part in response["content_parts"]:
        if part["type"] == "audio":
            path = Path(f"multimodal-output.{part['format']}")
            path.write_bytes(base64.b64decode(part["data"]))
            print("Wrote", path)

asyncio.run(main())

File: examples/multimodal_output.rs
//! Request text plus generated audio and save the audio output.

use base64::prelude::*;
use ferragent::llm::openai::OpenAiModel;
use ferragent::{Agent, ContentPart, Message};

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    dotenvy::dotenv().ok();
    let model = OpenAiModel::new("gpt-audio-1.5", std::env::var("OPENAI_API_KEY")?)
        .with_modalities(["text", "audio"])
        .with_audio_output("wav", "alloy");
    let mut agent = Agent::builder(model).build();

    let response = agent
        .run_message(Message::user(
            "Say a friendly two-sentence welcome to Rust.",
        ))
        .await?;
    if let Some(text) = response.content {
        println!("Transcript: {text}");
    }
    for part in response.content_parts {
        if let ContentPart::Audio { data, format, .. } = part {
            let path = format!("multimodal-output.{format}");
            std::fs::write(&path, BASE64_STANDARD.decode(data)?)?;
            println!("Wrote {path}");
        }
    }
    Ok(())
}
